### DISCUSSION TITLE
How to support CPU usage on Windows and Linux

### DISCUSSION THREAD

**Discussion by Wyrdgirn** [upscayl/upscayl#580] [Ideas] [answered]
### Checklist

- [X] I have checked that this issue isn't addressed in [the Wiki](https://github.com/upscayl/upscayl/wiki).
- [X] I have checked that this issue isn't addressed in any other issue or pull request.

### Description

The program does not work on machines without Vulkan, but it might work if emulated, so it would be very helpful if they added the option to force the use of swiftshader. This can be done by modifying Upscayl's files on Windows and Linux. On the former, it is in C:\Program Files\Upscayl by default.

I have tried deleting the **_vulkan-1.dll_** library and renaming a copy of **_vk_swiftshader.dll_** to **_vulkan-1.dll_**, which I put into the C:\Program Files\Upscayl\resources\bin folder. Then, set the GPU ID to 0 or leave it blank and the app works correctly. I have already tried all the models and they work (although slow) even on an Intel Atom N455 with 2gb of ram and Windows 10 🤯

So I would consider an option in settings (or automatically) to force CPU usage only for those systems without Vulkan useful... and even if it supports it it can become a quick fix for those with driver issues, etc. 🤔

Oh, you can also post in the documentation how to convert the app to CPU only by renaming and replacing the libraries as I mentioned before if the option is not useful for you, then at least more people will have access to this great tool 😎👌 You decide 😉

Greetings 😃✌️

Oh, by the way, excuse my English, my native language is Spanish 😅

### Related Problem(s)

The app does not work if the machine does not support Vulkan. Note that, as a workaround, there is still a possibility for this to not work.

### Alternatives

Renaming a copy of the **_vk_swiftshader.dll_** library to **_vulkan-1.dll_**

**Comment by aaronliu0130** [comment]
# 🤯

**Comment by NayamAmarshe** [comment]
I think we can add this info to the wiki.

**Comment by aaronliu0130** [comment]
I assume the same can be done on Linux by renaming libvk_swiftshader.so to libvulkan.so.1 . Not sure about Mac.

Maybe we should make a separate distribution for such builds?

**Comment by NayamAmarshe** [comment]
No, I think it'd be too much work for what's basically a smart hack.

I think it'd be better to let hacks stay in the wiki.

**Comment by NayamAmarshe** [comment]
Added here: https://github.com/upscayl/upscayl/wiki/Misc

**Comment by danielrode** [comment]
> I assume the same can be done on Linux by renaming libvk_swiftshader.so to libvulkan.so.1 . Not sure about Mac.
> 
> Maybe we should make a separate distribution for such builds?

When I try this on Void Linux using the AppImage version of Upscayl, I get an error:

_"BACKEND REPORTED:  image upscayl:  /tmp/.mount_upscay64N2F7/resources/bin/upscayl-bin: symbol lookup error: /tmp/.mount_upscay64N2F7/resources/bin/upscayl-bin: undefined symbol: vkAllocateDescriptorSets"_

**Reply by Wyrdgirn** [reply]
You can try this if you haven't tried it yet:

1) Reinstall Upscayl

And:

2) Delete or rename the file libvulkan.so

3) Create a symbolic link to libvk_swiftshader.so with the name libvulkan.so
`$ ln -s libvk_swiftshader.so /path/to/Upscayl/libvulkan.so`

Or:

2) Delete or rename the file libvulkan.so.1

3) Create a symbolic link to libvk_swiftshader.so with the name libvulkan.so.1
`$ ln -s libvk_swiftshader.so /path/to/Upscayl/libvulkan.so.1`

4) If libvulkan.so does not exist create a symbolic link to libvulkan.so.1 with the name libvulkan.so
`$ ln -s libvulkan.so.1 /path/to/Upscayl/libvulkan.so`

Try and comment 😉

**Comment by aaronliu0130** [comment]
Sounds like your driver really hates Vulkan. Sorry.

**Comment by CableSix** [comment]
I think a video teaching this would be good.

**Reply by aaronliu0130** [reply]
It's some files-with-the-same-titles' renamings; I don't really see the difficulty.

**Reply by CableSix** [reply]
It didn't work for me unfortunately.

**Comment by Dean-Corso** [comment]
Does not work for me (Windows) even with deleting the vulkan-1.dll & renaming the vk_swiftshader.dll vulkan-1.dll pop ups always the same error trying to upscale any image.
![Error_2024-01-27_005708](https://github.com/upscayl/upscayl/assets/88513086/99c3c595-a2b0-4111-8a31-d9c7b2fd5402)
A CPU support would be nice instead of just using GPU. Most other tools having these option the user can choose.

**Reply by aaronliu0130** [reply]
> A CPU support would be nice instead of just using GPU.

That would be impossible without rewriting everything and possibly switching to an entirely different method of upscaling.

**Reply by Wyrdgirn** [reply]
I think it is recommended that you use images in png format for both input and output, also, do not delete or rename the original vk_swiftshader.dll file, you must do it with a copy of it. If it still does not work then download another version of vulkan_swiftshader from the internet and replace the vulkan-1.dll file and try to see if it works for you. If it still doesn't work, uninstall and install the old v2.5.5 which I have verified works even on a notebook with Intel atom n455 + 2GB Ram + 2GB Pagefile + Windows 10 (one of the first versions of 2015...). I haven't used it in a while but I will soon test to see if the new versions still work in cpu mode...

**Comment by Dean-Corso** [comment]
Hi guys,

I did not delete any file just did rename it as you mention in your first post.
`Renaming the vk_swiftshader.dll library to vulkan-1.dll`
So what is the right way now? I tried it again using the original dll file names & files and just using png image for input & output but I still get the same error message. What version of vk_swiftshader.dll should I use? Just have 5.0.0.1 file version of it even from 2021 but they also fail on the same way. About v2.5.5, so there is no portable zip.

PS: "do not delete or rename the original vk_swiftshader.dll file, you must do it with a copy of it" = ? Why do you say that? You told us to rename the original **vk_swiftshader.dll** to **vulkan-1.dll**. Somehow you statement doesn't match. I just did this...
```
vulkan-1.dll		Renamed to _vulkan-1.dll
vk_swiftshader.dll	Renamed to vulkan-1.dll
```
...so if this was wrong then just tell us again the correctly way etc. Thanks.

**Reply by aaronliu0130** [reply]
Upscayl needs both files. It won't work with the swiftshader file missing.

**Reply by Dean-Corso** [reply]
```
vulkan-1.dll		Renamed to _vulkan-1.dll
vk_swiftshader.dll	Copy to vk_swiftshader_2.dll
vk_swiftshader.dll	Renamed to vulkan-1.dll
vk_swiftshader_2.dll	Renamed to vk_swiftshader.dll

In Folder now upscayl-2.9.8-win
-------------------------------------------------------
vulkan-1.dll		=	vk_swiftshader.dll
vk_swiftshader.dll	=	vk_swiftshader.dll
```
Also in this case when the original vk_swiftshader.dll is there and the vulkan-1.dll (which is vk_swiftshader.dll) it fails too with same error message. It's just not working. So can anyone test it whether it still works or not (Win 10 x64)?

**Reply by Wyrdgirn** [reply]
> You told us to rename the original vk_swiftshader.dll to vulkan-1.dll. Somehow you statement doesn't match...

Really?

> I have tried deleting the vulkan-1.dll library and renaming a **_copy_** of vk_swiftshader.dll to vulkan-1.dll

In the alternatives section I did not mention the original file, and in the entry I specified that it was made with a copy of swiftshader, not the original... Well, never mind, maybe it can be edited later to make it more specific, I guess.

Each update may use new vulkan functions that the included version of swiftshader may not support or support to a limited extent, which would cause execution problems, but in these cases using another version of the library would solve the problem.

Another thing to keep in mind is that the program will use even more Ram and CPU when running in an emulated vulkan environment, in such cases this method can cause problems on some potato/toaster machines, with 2gb of RAM I have had to use a 2gb page file to fill at least 4gb of virtual memory.

If it works or not, it would be great if you mentioned the brand of the cpu since I have only tested the hack on Intel with Windows 10 x64, This way we could keep track of which machines it worked on, if it only works with Intel or if on some machines it only works with a specific OS, I can't guarantee that it will work on AMD machines so...

Saludos / Greetings 😉✌️

**Comment by MohammedMehdiTBER** [comment]
It worked, but It is extremely slow. Tested in Arch Linux with Swift Shaders

**Comment by Wyrdgirn** [comment]
I have edited the post a little so that it is understood that it must be done with a copy and not with the original vk_swiftshader to avoid confusion. 😉✌️

**Comment by Wyrdgirn** [comment]
Upscayl v2.9.9 works on Intel Atom N455 and N550, 2GB Ram and 1GB of pagefile, the procedure is the same, slow with the gui minimized and very slow maximized, I would recommend an option to disable heavy animations in the gui to leave the biggest cpu load to the scaling process, idk, It's just a suggestion, minimized works but having it maximized with just the progress percentage updating every couple of minutes would be great...

Btw, I think you should update the readme.md file in the repository, in the current one you suggest a link to the old entry (issue), not to the discussion (that is, this one...), and I think it would be better to link it directly to the discussion since it is the most up-to-date... obviously the notice about the discussion appears but people will be guided first by the previous one before coming to this... Another option would be to update both entries to keep the information the same in both but I think I prefer the first option 😅

**Comment by wolli1234** [comment]
Does not seem to work on linux ubuntu with Upscayl 2.10.0 unfortunately

After a really short time I get a success message for upscaling, but no file is created

**Comment by NurudinImsirovic** [comment, marked as answer]
I think the wiki is a bit misguided. The vulkan file should not reside in the location of Upscayl.exe, but the actual upscaler executable, which is located at:
```
C:\Program Files\Upscayl\resources\bin
```

So what this means is you copy:
```
C:\Program Files\Upscayl\vk_swiftshader.dll
```
to:
```
C:\Program Files\Upscayl\resources\bin
```

and rename it to:
```
vulkan-1.dll
```

---

Verified that this works on the latest release:
![0](https://github.com/upscayl/upscayl/assets/102489121/86db875b-a8b6-4ba0-a551-debcf52583fd)

**Reply by Liuguoyong621** [reply]
Thank you very much, this way is OK, my computer CPU is AMD

**Reply by cbxbw** [reply]
This method does have an effect, but the running speed is too slow. The 2k image is enlarged by 3 times and the CPU runs at 80%. It takes about an hour to produce the image. My CPU is AMD 7840HS

**Reply by aaronliu0130** [reply]
@cbxbw That's as fast as it gets without a GPU.

**Reply by jarallahassalami** [reply]
thank you.
it is working

**Comment by MWMartley001** [comment]
I just downloaded the Windows package. It runs successfully however it doesn't store the image in the target destination. Any idea on this?

**Reply by MWMartley001** [reply]
Ok I just tried to use the same folder as the source file location and that worked; so there is a bug with saving to a different folder.

**Reply by aaronliu0130** [reply]
Could you report that by filing a bug report?

**Reply by MWMartley001** [reply]
Will do!

**Comment by djrickydadragon** [comment]
This method works for me. But because of going heavy on my CPU now, I found a 'per task cpu limiter' to reduce its demand a little. Now rendering will take a little longer but will safe the CPU from extra heath https://github.com/martin-kudlicka/cpu-limiter. 

You have to set a rule then to point it then to upscayl-bin.exe on the left and right , and then set the max of %. ( upscayl-bin.exe is only visible in taskmanager during rendering )

_upscayl-bin.exe in taskmanager_
![image](https://github.com/user-attachments/assets/559125c9-f4de-480d-91ec-d82c7228f931)

_example rule_ 
![image](https://github.com/user-attachments/assets/f6e36e60-e1d3-4673-a138-59eb98474773)

**Comment by dima333750** [comment]
`Renaming a copy of the vk_swiftshader.dll library to vulkan-1.dll`
It works on windows 10 + amd vega 8 
GPU is loaded by 50%

**Comment by dmlls** [comment]
Got it running on Arch Linux by installing SwiftShader (https://aur.archlinux.org/packages/swiftshader-git) and pointing the `VK_ICD_FILENAMES` environment variable to `/usr/share/vulkan/icd.d/vk_swiftshader_icd.json`, i.e., in your `.bashrc` / `.zshrc`:

```bash
export VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/vk_swiftshader_icd.json
```

**Comment by djdhemitz** [comment]
get error for me in laptop asus a455l with nvidia geforce 820m :

🔄 Resetting image paths
🚧 UPSCAYL_PROGRESS:  vkEnumeratePhysicalDevices failed -3
🚨 Error: Invalid GPU Device

anyone can help me?

**Reply by notlousybook** [reply]
my brother has the same problem

**Comment by iggykvnski** [comment]
Good news, at least for Windows users with only CPU.

https://forums.anandtech.com/threads/upscayl-cpu-benchmark-much-improved-with-recompiled-dll.2631914/

It's much faster than doing the file swap of the original Vulkan SwiftShader files.

**Comment by iggykvnski** [comment]
Sorry. I don't understand. Not a Linux person.

What is being discussed? What is the value 2048 and 1024?

**Reply by dima333750** [reply]
I'm sorry, it's not a mistake here)
a lot of tabs are confused
